tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,211 @@
+---
+title: Type annotations in Python
+published: 2019-08-01
+tags: Python, typing, types, tdd
+language: python
+author-name: Carlos Villavicencio
+twitter-profile: po5i
+github-profile: po5i
+description: Learn how to improve the software quality and readability in our Python code by using the built-in typing system as it complements other software development processes like type-driven development and test-driven development.
+---
+
+_Type-Driven Development_ is a technique in which we plan our program development by adding types first and then write the code that satisfies that type. It brings many benefits such as increasing the robustness, accuracy, testability, readability, and extensibility of your code by taking advantage of the language’s type system. We strongly suggest as follow up reading [_Type-Driven Development with Idris_][tdd-idris] and extrapolate these practices to other languages.
+
+In a [previous article][types-vs-tests], we discussed the benefits of working with _Type-Driven Development_ and the importance of static type checking at compilation time for strongly typed languages such as Java or Haskell. Type-checking strengthens our code, reduces the number of tests we need to write which decreases our dependency on them no matter if the paradigm is object-oriented or functional
+
+## What about Python?
+
+Is it possible to use a strong type system in a language like Python? The answer is yes. [PEP-484][pep-484] introduced _Type Annotations to Python_ 3.5 back in 2014. It’s been a while, right? Why is this not popular? It's because Python being a dynamic language doesn’t require to explicitly write variable or return types, they’re completely optional and mostly used by editors and IDEs.
+
+In this article, we’ll be walking through the Python - how to add types to our code and Docstrings and perform static type checking using [mypy][mypy]. There are plenty options for static type-checking in Python, but in this tutorial we will follow [Guido van Rossum suggestion][guido-mypy].
+
+### Mypy
+
+Mypy is a third-party Python library that provides optional static type checking. Unlike other non-dynamic programming languages like Java, where the static type-checking takes place at compilation time, Mypy CLI does the type-check to a file (or a set of files), so it's up to you to decide where in your workflow the type-checking will be executed. It can be on CI or a pre-commit hook, just to mention some.
+
+To start using mypy, install it using any version of `pip` globally or in your virtual environment.
+
+```bash
+pip install mypy
+```
+
+To perform static type checking on a source file:
+
+```bash
+mypy my_file.py
+```
+
+## Primitive types
+
+Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value. This might be confusing, but it doesn't mean that Python has _weak typing_. Python is _strongly typed_. Surprised? Check this out:
```

Comment:
Was the conversation related to this resolved? Maybe adding a short comment about the difference between weak and strong typing could help to know what the example below is saying. Or a link to something with additional information.